Support Avro's decimal

aymkhalil opened the pull request:
This patch adds support for an AVRO only type: [decimal](https://avro.apache.org/docs/1.10.2/spec.html#Decimal).

It would enable conversion from other types like CQL Decimal available in C* CDC.

For example, the use can use an ARRO decimal aware sink with an upstream topic that has a CQL Decimal type like the following:
```JSON
{
  "name": "longitude",
  "type": [
    "null",
    {
      "type": "record",
      "name": "cql_decimal",
      "namespace": "",
      "fields": [
        {
          "name": "bigint",
          "type": "bytes"
        },
        {
          "name": "scale",
          "type": "int"
        }
      ],
      "logicalType": "cql_decimal"
    }
  ]
}
```

can be converted to Avro's decimal with the following converion:
```JSON
{
    "steps": [
        {
            "type": "compute",
            "fields": [
                {
                    "name": "value.longitude",
                    "expression": "fn:decimal(value.longitude.bigint, value.longitude.scale)",
                    "type": "DECIMAL"
                }
            ]
        }
    ]
}
``` 

The output schema of the conversion would look like this:
```JSON
{
  "type": "bytes",
  "logicalType": "decimal",
  "precision": 4, // would match the BigDecimal.precision that is retrievable after building the BigDecimal the takes bytes + scale as input (but not precision)
  "scale": 2 // would match the scale in the CQL decimal type
}
```

In order to support the conversion, a new function is registered with the following sig:
```
decimal(byte[] bytes, int scale) --> BigDecimal
```

eolivelli reviewed (commented):
Good idea

I wonder if it would work better 'toDecimal' instead of 'decimal'


a second thought:
probably the table has many columns of this type,  probably we could have ALSO a specific step 'convert Cassandra CDC logical types' that does the conversion for all the fields.

it will help a lot because it would be a no Brainerd and a simple checkbox in the UI

lhotari reviewed (approved):
LGTM. Amazing work @aymkhalil !

cbornet commented on README.md:
```diff
@@ -291,6 +292,9 @@ The Expression Language supports the following functions:
   * `input` a timestamp to add to.
   * `delta` a `long` amount of `unit` to add to `input`. Can be a negative value to perform subtraction.
   * `unit` the string unit of time to add or subtract. Can be one of [`years`, `months`, `days`, `hours`, `minutes`, `seconds`, `millis`].
+* `decimal(input, scale)`: Converts `input` to a `BigDecimal` with the given `scale`. 
```
Add the possibility to define the precision ?

aymkhalil replied:
I thought about this, I can add it as optional, but in practice it is only used for Avro's validation and when constructing java's BigDecimal, you can't specify precision - it is effectively the is the number of digits in the unscaled value so with bytes + scale only a user would have full control on their decimal.

cbornet replied:
nm, you're right. Adding precision is useless here

cbornet commented on tests/src/test/java/com/datastax/oss/pulsar/functions/transforms/tests/AbstractDockerTest.java:
```diff
@@ -389,6 +403,45 @@ public void testComputeAvroDateToString() throws Exception {
+        ("{\"steps\": [{\"type\": \"compute\", \"fields\": ["
+            //+ "{\"name\": \"value.cqlDecimal\", \"expression\": \"fn:decimal(value.cqlDecimal.bigint, value.cqlDecimal.scale)\", \"type\": \"DECIMAL\"},"
+            + "{\"name\": \"value.cqlDecimalOptional\", \"expression\": \"fn:decimal(value.cqlDecimalOptional.bigint, value.cqlDecimalOptional.scale)\", \"type\": \"DECIMAL\"},"
+            //+ "{\"name\": \"value.cqlDecimalAsString\", \"expression\": \"fn:decimal(value.cqlDecimal.bigint, value.cqlDecimal.scale)\", \"type\": \"STRING\"},"
```
```suggestion
```

cbornet commented on tests/src/test/java/com/datastax/oss/pulsar/functions/transforms/tests/AbstractDockerTest.java:
```diff
@@ -389,6 +403,45 @@ public void testComputeAvroDateToString() throws Exception {
+        org.apache.avro.Schema.createRecord("cql_decimal", "", "ns", false, fields);
+    org.apache.avro.generic.GenericRecord record = new GenericData.Record(avroSchema);
+
+    //BigDecimal decimal = new BigDecimal("123456789012345678901234567890.123456789");
```
```suggestion
```

cbornet commented on pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/ComputeStep.java:
```diff
@@ -188,29 +190,25 @@ private GenericRecord computeFields(
+      Object value = field.getEvaluator().evaluate(context);
+      ComputeFieldType type = field.getType() == null ? getFieldType(value) : field.getType();
+      Schema.Field avroField;
+      if (type == ComputeFieldType.DECIMAL) {
```
Can we put this logic in `createAvroField(field, type, value)` ?

aymkhalil replied:
sure

btw DECIMAL is the first type I encountered that the schema definition itself relies on the value/other metadata hence the special handing (and the lazy evaluation of schema fields after evaluating the values themselves).

cbornet commented on pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlTypeConverter.java:
```diff
@@ -450,6 +461,20 @@ protected OffsetDateTime coerceToOffsetDateTime(Object value) {
         MessageFactory.get("error.convert", value, value.getClass(), OffsetDateTime.class));
   }
 
+  protected BigDecimal coerceToBigDecimal(Object value) {
```
Do we need to have converters for numeric types ?
Also should other numeric types be able to coerce from BigDecimal ?

aymkhalil replied:
I could only think of String because the number + scale are self contained and gives user flexibility depending on how the value is generate in their upstream.

If it makes sense, I could coerce byte[] as well assuming it is meant to represent unscaled decimal (i.e. scale 0). I can't think of other meaningful conversions.

cbornet replied:
It's easy to create a BigDecimal from a number:
BigDecimal.valueOf(Long)
new BigDecimal(Double)

For the conversion from bytes, I think it's not clear that the bytes are representing a BigInteger (and not a Long or a Double). So I think we shouldn't support this conversion..

aymkhalil replied:
Ok, will do

aymkhalil commented on tests/src/test/java/com/datastax/oss/pulsar/functions/transforms/tests/AbstractDockerTest.java:
```diff
@@ -389,6 +403,45 @@ public void testComputeAvroDateToString() throws Exception {
+  public void testComputeCqlDecimalToAvroDecimal() throws Exception {
+    String userConfig =
+        ("{\"steps\": [{\"type\": \"compute\", \"fields\": ["
+            //+ "{\"name\": \"value.cqlDecimal\", \"expression\": \"fn:decimal(value.cqlDecimal.bigint, value.cqlDecimal.scale)\", \"type\": \"DECIMAL\"},"
```
Yeas I'll probably delete, or re-enable. If you check the TODO message above.

I was not sure why if I have both optional and non-optional decimal, publishing to the data topic would fail. I'll troubleshoot more. Perhaps, I can create to separate tests. The reason I wanted to test both is because the CQL Decimal will be non-optional in the key part (primary key, assuming it originated from C* CDC) but optional in the value.

aymkhalil replied:
I ended up using a KeyValue schema with non optional decimal in key, and optional in value and it went through (also it is closer to what CDC would do). The fact the both cannot coexist in the value remains a puzzle to me (seems logic in Avro and not pulsar client that would cause `java.io.EOFException`)

aymkhalil commented:
> I wonder if it would work better 'toDecimal' instead of 'decimal'

Actually I originally had it `toBigDecimal` (but `toDecimal` is a better one) - then I realized we had `toString` that we expose to user as `str` and we don't have `to*` convention so I thought `decimal` is more inline with the convention. I'm open to both 
 
> a second thought: probably the table has many columns of this type, probably we could have ALSO a specific step 'convert Cassandra CDC logical types' that does the conversion for all the fields.
> it will help a lot because it would be a no Brainerd and a simple checkbox in the UI

Yeah you are right this conversion could be reused multiple resulting in duplicate steps. It is like a "reasonable" work around for the CDC issue in specific. I was avoiding adding too specific of a conversion step to keep the transformation generic enough. I actually thought of adding another step like `compute` (but called `computeSchema`) that would define a schema conversion rule and is not specific to a particular field. For example:
```
{
    "steps": [
        {
            "type": "schema", // indicates that this step is a schema transformation step. Could be named "computeSchema" to draw analogy with the compute step 
            "fields": [
                {
                    "name": "cql_decimal", // the name of the schema to be transformed
                    "type": "record", // type of the schema. "record" it indicates the fields names are nested
                    "output": { // the desired output schema, could be named outputSchema or newSchema to be more explicit 
                        "type": "bytes", 
                        "value": "bigint", // another way to referece those is to use dot notations like cql_decimal.bigint or record.bigint. This would leverage the Expression Language for completeness
                        "logicalType": "decimal",
                        "precision": 4, // use can hard code stuff directly 
                        "scale": "scale" // again, could be cq_decimal.scale or record.scale
                    }
                }
            ]
        }
    ]
}
```
but it turned out more complex that I though, however I can pursue this direction if we see value in it

@cbornet what are your thoughts regarding both points above?

aymkhalil commented:
I realized that in practice, BigDecimal would be of variable precision/scale. So what will happen in this patch is if scale changes and we reuse the schema, validation could pass if no rounding is required: https://github.com/apache/avro/blob/master/lang/java/avro/src/main/java/org/apache/avro/Conversions.java#L122-L132, if we cache per field, then compatibility rules would apply which makes me thing this logical type might not be of great use after all

cbornet commented on pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlEvaluator.java:
```diff
@@ -81,6 +81,12 @@ private void registerFunctions() {
+        .mapFunction(
+            "fn",
+            "decimal",
+            JstlFunctions.class.getMethod("toBigDecimal", byte[].class, int.class));
```
We should implement coercion to BigInteger so you can take Object as first arg.
Also we should use Object for the second arg and coerce to Integer.
And to be complete it would be nice to have toBigDecimal(Object) that coerces the first arg to Double.

cbornet commented on pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlTypeConverter.java:
```diff
@@ -239,6 +245,12 @@ protected byte[] coerceToBytes(Object value) {
     if (value instanceof byte[]) {
       return (byte[]) value;
     }
+    if (value instanceof ByteBuffer) {
```
When do we get ByteBuffer ?

aymkhalil replied:
It comes from the custom CDC CQL Type (the bigint part [is encoded as ByteBuffer](https://github.com/datastax/cdc-apache-cassandra/blob/cd64bdd03af7a608687b6c54aa614021c62d8027/commons/src/main/java/com/datastax/oss/cdc/CqlLogicalTypes.java#L156)) so I mimicked the same in the tests. The `coerceToBytes` from ByteBuffer will kick in when expression like `value.cqlDecimalField.bigint` are evaluated

cbornet replied:
So this is only for tests ? Can we use instead the type that we'll get from the AVRO deserializer (probably byte[] ?)

aymkhalil replied:
Not only for testing. If the byte[] was encoded as ByteBuffer, the AVRO deserializer will return ByteBuffer even if the schema type is bytes. 

This how CDC code converts it: https://github.com/datastax/cdc-apache-cassandra/blob/cd64bdd03af7a608687b6c54aa614021c62d8027/commons/src/main/java/com/datastax/oss/cdc/CqlLogicalTypes.java#L146

cbornet replied:
I don't see how the info of which structure was used byte[]/ByteBuffer during serialization would be in the AVRO record...

aymkhalil replied:
I just realized the conversation from ByteBuffer had been removed from the `coerceToBytes` and is now limited to `coerceToBigInteger` only. However, I'm not sure how the AVRO recognizes ByteBuffer under the hoods.

cbornet commented on pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlTypeConverter.java:
```diff
@@ -450,6 +462,23 @@ protected OffsetDateTime coerceToOffsetDateTime(Object value) {
+    throw new ELException(
+        MessageFactory.get("error.convert", value, value.getClass(), BigDecimal.class));
+  }
+
```
We should add the coercions from BigDecimal to String, Long, Double, etc...

cbornet commented on tests/src/test/java/com/datastax/oss/pulsar/functions/transforms/tests/util/CqlLogicalTypes.java:
```diff
@@ -1,9 +1,27 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
The copyright header is there twice

cbornet commented on tests/src/test/java/com/datastax/oss/pulsar/functions/transforms/tests/util/CqlLogicalTypes.java:
```diff
@@ -1,9 +1,27 @@
+ */
 package com.datastax.oss.pulsar.functions.transforms.tests.util;
 
+import org.apache.avro.Conversion;
```
Unused ?

cbornet commented on tests/src/test/java/com/datastax/oss/pulsar/functions/transforms/tests/util/CqlLogicalTypes.java:
```diff
@@ -1,9 +1,27 @@
 import org.apache.avro.Schema;
 import org.apache.avro.SchemaBuilder;
 import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecordBuilder;
+import org.apache.avro.generic.IndexedRecord;
```
Unused ?

cbornet commented on tests/src/test/java/com/datastax/oss/pulsar/functions/transforms/tests/util/CqlLogicalTypes.java:
```diff
@@ -22,6 +40,7 @@
  */
 
 import java.math.BigDecimal;
+import java.math.BigInteger;
```
Unused ?

aymkhalil commented on pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlEvaluator.java:
```diff
@@ -81,6 +81,19 @@ private void registerFunctions() {
+            .getFunctionMapper()
+            .mapFunction(
+                    "fn",
+                    "decimalFromDouble",
```
EL doesn't support methods overloads, other names/suggestions are welcome (maybe `decimal(double, null)`? Because the coercion rules of the first arg will depend on the second, I thought it could be confusing to understand the behavior, hence separate methods).

aymkhalil replied:
or maybe `decimalFromNumber`

cbornet replied:
Yes, decimalFromNumber is better.
Sorry, only saw this now.

aymkhalil commented:
@cbornet I addressed your feedback, PTAL 
However, I think the DECMAL type (although compliant with AVRO) is not practical, here is why in context of CDC:
```CQL
create table ks1.d2 (id int primary key, v1 decimal) with cdc=true;
insert into ks1.d2(id, v1) values (1, 1.334);
``` 
assume the following transform is in place:
```JSON
{
   "steps":[
      {
         "type":"compute",
         "fields":[
            {
               "name":"value.v2",
               "expression":"fn:decimal(value.v1.bigint, value.v1.scale)",
               "type":"DECIMAL"
            }
         ]
      }
   ]
}
```
works fine, results in the following schema:
```JSON
{
        "name": "v2",
        "type": [
          "null",
          {
            "type": "bytes",
            "logicalType": "decimal",
            "precision": 4,
            "scale": 3
          }
        ]
      }
```

now decreasing the scale works fine as well (without changing the schema on the topic):
```CQL
 insert into ks1.d2(id, v1) values (1, 1.3);
```

however, increasing the scale is problematic:
```CQL
insert into ks1.d2(id, v1) values (1, 1.33499);
```

will fail with
```
org.apache.avro.AvroTypeException: Cannot encode decimal with scale 5 as scale 3 without rounding in field v2
```

so basically depending on the very first decimal that was received, the scale will limit all future decimal values which is impractical. Otherwise, this patch is technically correct. Let me know WDYT.

cbornet commented:
> so basically depending on the very first decimal that was received, the scale will limit all future decimal values which is impractical.

Is this bc it sets the schema on the topic with the first scale received ?

cbornet commented on README.md:
```diff
@@ -291,7 +292,11 @@ The Expression Language supports the following functions:
+* `decimal(input, scale)`: Converts `input` to a `BigDecimal` with the given `scale`. 
+  * `input` unscaled value of the BigDecimal. Can be any of STRING, INTEGER, LONG or Array of bytes containing the two's-complement representation in big-endian byte order.
+  * `scale` the scale of the `BigDecimal` to create.
+* `decimalFromDouble(input)`: Converts `input` to a `BigDecimal`.
```
Would call it `decimalFromNumber`. Sorry for the late notice.

cbornet commented on README.md:
```diff
@@ -291,7 +292,11 @@ The Expression Language supports the following functions:
   * `delta` a `long` amount of `unit` to add to `input`. Can be a negative value to perform subtraction.
   * `unit` the string unit of time to add or subtract. Can be one of [`years`, `months`, `days`, `hours`, `minutes`, `seconds`, `millis`].
-
+* `decimal(input, scale)`: Converts `input` to a `BigDecimal` with the given `scale`. 
```
Call it decimalFromUnscaled ?

cbornet commented:
LGTM.
We can rebase and merge.
As discussed it would be useful to have a decimalFromNumberWithScale(decimal, scale) to be able to fix the scale of the output in CDC but it can be done in a subsequent PR.

aymkhalil commented:
> LGTM. We can rebase and merge. As discussed it would be useful to have a decimalFromNumberWithScale(decimal, scale) to be able to fix the scale of the output in CDC but it can be done in a subsequent PR.

Issue for tracking: https://github.com/datastax/pulsar-transformations/issues/75
